ratelimits/limit.go

```diff
@@ -0,0 +1,142 @@
+	// allowed. It must be greater than zero.
+	Period config.Duration
+}
+
+func validateLimit(l limit) error {
+	if l.Burst <= 0 {
+		return fmt.Errorf("invalid burst '%d', must be > 0", l.Burst)
+	}
+	if l.Count <= 0 {
+		return fmt.Errorf("invalid count '%d', must be > 0", l.Count)
+	}
+	if l.Period.Duration <= 0 {
+		return fmt.Errorf("invalid period '%s', must be > 0", l.Period)
+	}
+	return nil
+}
+
+type limits map[string]limit
+
+// loadLimits marshals the YAML file at path into a map of limis.
+func loadLimits(path string) (limits, error) {
+	lm := make(limits)
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	err = strictyaml.Unmarshal(data, &lm)
+	if err != nil {
+		return nil, err
+	}
+	return lm, nil
+}
+
+// parseOverrideNameId is broken out for ease of testing.
+func parseOverrideNameId(key string) (*Name, string, error) {
+	if !strings.Contains(key, ":") {
+		// Avoids a potential panic in strings.SplitN below.
+		return nil, "", fmt.Errorf("invalid override %q, must be formatted 'name:id'", key)
+	}
+	nameAndId := strings.SplitN(key, ":", 2)
+	nameStr := nameAndId[0]
+	if nameStr == "" {
+		return nil, "", fmt.Errorf("empty name in override %q, must be formatted 'name:id'", key)
+	}
+
+	name, ok := stringToName[nameStr]
+	if !ok {
+		return nil, "", fmt.Errorf("unrecognized name %q in override limit %q, must be one of %v", nameStr, key, limitNames)
+	}
+	id := nameAndId[1]
+	if id == "" {
+		return nil, "", fmt.Errorf("empty id in override %q, must be formatted 'name:id'", key)
+	}
+	return &name, id, nil
+}
+
+// loadAndParseOverrideLimits loads override limits from YAML, validates them,
+// and parses them into a map of limits keyed by 'Name:id'.
+func loadAndParseOverrideLimits(path string) (limits, error) {
+	fromFile, err := loadLimits(path)
+	if err != nil {
+		return nil, err
+	}
+	parsed := make(limits, len(fromFile))
+
+	for k, v := range fromFile {
+		err = validateLimit(v)
+		if err != nil {
+			return nil, fmt.Errorf("validating override limit %q: %w", k, err)
+		}
+		name, id, err := parseOverrideNameId(k)
+		if err != nil {
+			return nil, fmt.Errorf("parsing override limit %q: %w", k, err)
+		}
+		err = validateIdForName(*name, id)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"validating name %s and id %q for override limit %q: %w", nameToString[*name], id, k, err)
+		}
+		if *name == CertificatesPerFQDNSetPerAccount {
+			// FQDNSet hashes are not a nice thing to ask for in a config file,
+			// so we allow the user to specify a comma-separated list of FQDNs
+			// and compute the hash here.
+			regIdDomains := strings.SplitN(id, ":", 2)
+			if len(regIdDomains) != 2 {
+				// Should never happen, the Id format was validated above.
+				return nil, fmt.Errorf("invalid override limit %q, must be formatted 'name:id'", k)
+			}
+			regId := regIdDomains[0]
+			domains := strings.Split(regIdDomains[1], ",")
+			fqdnSet := sa.HashNames(domains)
```

Comment:
Kinda unfortunate that we haul in the whole SA package for this one function. Can we reduce dependency this by moving HashNames elsewhere?

Reply:
I'm not sure I understand. Boulder ships as a single binary, what is the benefit of breaking HashNames out of the SA package? Do we expect outside orgs to import this package?

Reply:
They've done similar things for other packages! See https://github.com/letsencrypt/boulder/issues/6910 for an example. Also, I'd still love to break the boulder binary back up at some point -- I don't want CA code running on RVAs, and I really don't want WFE code running on CAs -- so relying on that bundling makes me a little bit sad.

Reply:
I've gone ahead and moved it to the core package as that seems to be the most appropriate place.